File: Medium/LongestSubstringWithoutRepeatingCharacters.py
__author__ = 'Lei Chen'
'''
Given a string, find the length of the longest substring without repeating characters.
For example, the longest substring without repeating letters for "abcabcbb" is "abc",
which the length is 3. For "bbbbb" the longest substring is "b", with the length of 1
'''

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s: return 0
        if len(s) <= 1: return len(s)

        prefixStart = 0
        prefixStop = 0
        longest = 1
        for i, c in enumerate(s[1:],start=1):
            if c not in s[prefixStart:prefixStop+1]:
                prefixStop = i
                longest = max(longest, prefixStop - prefixStart + 1)

            else:
                longest = max(longest, prefixStop - prefixStart + 1)
                index = s[prefixStart:prefixStop+1].index(c)
                prefixStart = prefixStart+index+1
                prefixStop = i

        return longest



s = Solution()
print(s.lengthOfLongestSubstring("abcabcbc"))

File: java/LongestSubstringWithoutRepeatingCharacters.java
public class LongestSubstringWithoutRepeatingCharacters {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		LongestSubstringWithoutRepeatingCharacters s = new LongestSubstringWithoutRepeatingCharacters();
		System.out.println(s.lengthOfLongestSubstring("abcabcbb"));
		System.out.println(s.lengthOfLongestSubstring("bbbb"));
	}

	/*
	 * Given a string, find the length of the longest substring without repeating characters. For example, the longest
	 * substring without repeating letters for "abcabcbb" is "abc", which the length is 3. For "bbbbb" the longest
	 * substring is "b", with the length of 1.
	 */
	public int lengthOfLongestSubstring(String s) {
		if (null == s || s.length() <= 0)
			return 0;

		int charSetSize = 255;
		char[] arr = s.toCharArray();
		int size = arr.length;

		int left = 0;
		int right = 0;
		int length = 1;

		int maxLength = 1;

		int[] charsPos = new int[charSetSize];
		for (int i = 0; i < charSetSize; i++) {
			charsPos[i] = -1;
		}

		for (int i = 0; i < size; i++) {
			char c = arr[i];
			int index = c;
			if (charsPos[index] >= 0) { // char in prefix
				// move the left to first old char's right
				int nextLeft = charsPos[index] + 1;
				for (int j = left; j < nextLeft; j++) {
					char oldC = arr[j];
					int oldIndex = oldC;
					charsPos[oldIndex] = -1;
				}

				left = nextLeft;
			}

			right = i;
			length = right - left + 1;
			charsPos[index] = i;

			if (length > maxLength) {
				maxLength = length;
			}
		}

		return maxLength;

	}
}